[FINAL] feat: [EXC-1676] add allowed viewers variant to canister's log visibility (#326)
dfinity/interface-spec: spec/index.md

The total size of all returned logs does not exceed 4KiB.
If new logs are added resulting in exceeding the maximum total log size of 4KiB, the oldest logs will be removed.
Logs persist across canister upgrades and they are deleted if the canister is reinstalled or uninstalled.

The log visibility is defined in the `log_visibility` field of `canister_settings` and it can be:

- `controllers`: only visible to the canister's controllers (by default)

> - `allowlist`: visible to the provided list of principals, max number is limited by 10

Reviewer: ```suggestion
- `allowlist`: visible to the provided list of principals, the maximum length of the list is 10
```

Author: done.

> - `allowed_viewers` (`vec principal`): only the provided list of principals can fetch logs, the maximum length of the list is 10

Reviewer 2: ```suggestion
- `allowed_viewers` (`vec principal`): only the provided list of principals and the canister's controllers can fetch logs, the maximum length of the list is 10
```

If new logs are added resulting in exceeding the maximum total log size of 4KiB, the oldest logs will be removed.
Logs persist across canister upgrades and they are deleted if the canister is reinstalled or uninstalled.

The log visibility is defined in the `log_visibility` field of `canister_settings`:

- `controllers`: only canister's controllers can fetch logs (by default)

> - `allowed_viewers`: only provided list of principals can fetch logs, the maximum length of the list is 10

Reviewer: ```suggestion
- `allowed_viewers`: only the provided list of principals can fetch logs, the maximum length of the list is 10
```

Author: done.

}
CanisterLogVisibility
  = Controllers

>   | AllowedViewers

Reviewer: ```suggestion
  | AllowedViewers [Principal]
```

Author: Done.

  certified_data: CanisterId ↦ Blob;
  canister_history: CanisterId ↦ CanisterHistory;
  canister_log_visibility: CanisterId ↦ CanisterLogVisibility;

>   canister_log_allowlist: CanisterId ↦ Set Principal;

Reviewer: I wouldn't introduce a new mapping `canister_log_allowlist` here since `canister_log_visibility` contains all the necessary information. This suggestion then also makes the question how to update this mapping obsolete. Please see my suggestion below on how to formulate the condition if a request sender is authorized to fetch the canister logs.

Author: Ok, removed `canister_log_allowlist`.

if A.settings.log_visibility is not null:
  New_canister_log_visibility = A.settings.log_visibility
  if New_canister_log_visibility is AllowList:

>     New_canister_log_allowlist = A.settings.log_visibility ?? how to get allowlist?

Author: @mraszyk could you please suggest on how to write down updating allowlist in this notation?

Author: Removed `New_canister_log_allowlist`.

Q.method_name = 'fetch_canister_logs'
Q.arg = candid(A)
A.canister_id = effective_canister_id
(S[A.canister_id].canister_log_visibility = Public)
  or 
  (S[A.canister_id].canister_log_visibility = Controllers and Q.sender in S[A.canister_id].controllers)
  or 

>   (S[A.canister_id].canister_log_visibility = AllowedViewers Principals and Q.sender in Principals) 

Reviewer: ```suggestion
  (S[A.canister_id].canister_log_visibility = AllowedViewers Principals and (Q.sender in S[A.canister_id].controllers or Q.sender in Principals))
```
to always allow controllers to fetch the logs.

Q.method_name = 'fetch_canister_logs'
Q.arg = candid(A)
A.canister_id = effective_canister_id

> S[A.canister_id].canister_log_visibility = 
>   Public 

Reviewer: ```suggestion
(S[A.canister_id].canister_log_visibility = Public)
```
it's a top-level disjuction that we're splitting across lines

Author: done.

Q.method_name = 'fetch_canister_logs'
Q.arg = candid(A)
A.canister_id = effective_canister_id

> S[A.canister_id].canister_log_visibility = Public or (S[A.canister_id].canister_log_visibility = AllowedViewers Principals and Q.sender in Principals) or Q.sender in S[A.canister_id].controllers

Reviewer: ```suggestion
S[A.canister_id].canister_log_visibility = Public or (S[A.canister_id].canister_log_visibility = AllowedViewers Principals and Q.sender in Principals) or (S[A.canister_id].canister_log_visibility = Controllers and Q.sender in S[A.canister_id].controllers)
```
or can controllers always fetch the logs even if the log visibility is AllowedViewers and the controllers are not on the list?

Author: shall we allow developers to forbid fetching the logs? if yes, then allowed_viewers can be empty and should not include controllers.

Reviewer: > shall we allow developers to forbid fetching the logs?

that sounds like a good motivation to apply my suggestion then; hopefully there aren't many cases when a canister has many controllers and only needs many additional log viewers (as there can be at most 10 allowed viewers, but in such a case we can bump the limit of 10 I suppose)

Reviewer: > shall we allow developers to forbid fetching the logs?

We discussed in the spec meeting and in fact, one can't really prevent canister controllers from fetching the logs as they can just override the settings, fetch the logs, and set the allowed viewers back to the original list of principals. Moreover, when the set of controllers changes, the change needs to be synced in both the list of controllers and the list of allowed log viewers. So maybe it's worthwhile reconsidering the semantics of the allowed viewers' list w.r.t. canister controllers.

Reviewer 3: @mraszyk Are you then suggesting that the semantics should be something like:

1. private -> only controllers can view
2. allowed_viewers -> controllers + principals in this list can view
3. public -> any principal is allowed to view

Q.method_name = 'fetch_canister_logs'
Q.arg = candid(A)
A.canister_id = effective_canister_id

> S[A.canister_id].canister_log_visibility = Public or Q.sender in S[A.canister_id].controllers or Q.sender in S[A.canister_id].canister_log_allowlist

Reviewer: ```suggestion
S[A.canister_id].canister_log_visibility = Public or (S[A.canister_id].canister_log_visibility = allowlist Principals and Q.sender in Principals) or Q.sender in S[A.canister_id].controllers
```

Author: Done.
